keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -0,0 +1,617 @@
+  &metrics.HistogramOpts{
+	  Subsystem: Subsystem,
+    Name: "endpoint_overload_by_zone",
+    Help: "Overload for endpoints by zone on each Service sync",
+  },
+  []string{"zone"}, // zone name
+)
+
+// This metric observes churn of EndpointSlices per sync
+EPSChangedPerSync = metrics.NewHistogramVec(
+  &metrics.HistogramOpts{
+	  Subsystem: Subsystem,
+    Name: "endpointslices_changed_per_sync",
+    Help: "Number of EndpointSlices be changed on each Service sync",
+  },
+  []string{"approach"}, // either "random" or "auto"
+)
+
+// EndpointSliceSyncs tracks the number of sync operations the controller runs along with their result.
+EndpointSliceSyncs = metrics.NewCounterVec(
+  &metrics.CounterOpts{
+    Subsystem:      EndpointSliceSubsystem,
+    Name:           "syncs",
+    Help:           "Number of EndpointSlice syncs",
+    StabilityLevel: metrics.ALPHA,
+  },
+  []string{"result"}, // either "success" or "failure"
+)
+
+```
+
+### Graduation Criteria
+- Alpha should provide basic functionality covered with tests described above.
+
+### Version Skew Strategy
+This KEP requires updates to both the EndpointSlice Controller and kube-proxy.
+Thus there could be two potential version skew scenarios:
+1. EndpointSlice Controller falls back to current behavior that does not
+   support labeling EndpointSlices. In this case, kube-proxy will still work
+   because EndpointSlices will not include topology hints.
+2. Kube-Proxy falls back to current behavior that does not support topology
+   hints in EndpointSlices. In this case, kube-proxy will continue to consume
+   all endpoints. This will not be an issue, it simply won't be taking advantage
+   of the new controller functionality.
+
+Each scenario described above will end up behaving as if this feature is not
+enabled even if the Service annotation has been set.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: TopologyAwareHints
+    - Components depending on the feature gate:
+      - kube-controller-manager
+      - kube-proxy
+
+* **Does enabling the feature change any default behavior?**
+  No.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+  Yes. It can easily be disabled universally by turning off the feature gate or
+  on a specific Service by setting the annotation to `Disable`.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+  EndpointSlices hints will be added again resulting in changes to existing
+  EndpointSlices for Services that have this feature enabled.
+
+* **Are there any tests for feature enablement/disablement?**
+  This feature is not yet implemented but per-Service enablement/disablement is
+  covered in depth as part of the test plan.
+
+### Rollout, Upgrade and Rollback Planning
```

Comment:
Technically neither of the below is needed for Alpha, but I'm really happy to see that already filled in :)